Field instance cannot expand to another feature on the checked tile, if it's not on the board

### Description
This TODO needs to be addressed:
https://github.com/YetAnotherSpieskowcy/Carcassonne-Engine/blob/a711a81f37b51efd849b63710c001a60ab8971fb/pkg/game/board.go#L240-L243

### Graphical example 
![image](https://github.com/user-attachments/assets/ead15437-4b8a-4642-9918-8250173ad766)

### Regression test
Already present in the affected branch (#62), you'll just need to remove the `t.Skip` to have it run:
https://github.com/YetAnotherSpieskowcy/Carcassonne-Engine/blob/a711a81f37b51efd849b63710c001a60ab8971fb/pkg/game/board_test.go#L137-L177

### Test output
```
$ go test -count=1 ./pkg/game -run TestBoardFieldCanBePlacedReturnsFalseWhenExpandToFieldWithMeepleHappensOverAnotherField
--- FAIL: TestBoardFieldCanBePlacedReturnsFalseWhenExpandToFieldWithMeepleHappensOverAnotherField (0.00s)
    board_test.go:173: expected false, got true instead
FAIL
FAIL	github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/game	0.002s
FAIL
```

--- a/pkg/game/board.go
+++ b/pkg/game/board.go
@@ -246,12 +246,6 @@ func (board *board) fieldCanBePlaced(tile elements.PlacedTile, feat elements.Pla
 	// This means we don't have to care about whether our field expands into
 	// a different feature on our tile - we will only expand the feature
 	// if it has a meeple and that only happens once.
-
-	// unset the Meeple in our copy of the feature
-	feat.Meeple = elements.Meeple{}
-
-	// TODO: Field instance cannot expand to another feature on the tile checked by
-	// this method since it's not part of the board. See GH-86.
 	field := field.New(feat, tile)
 
 	return field.IsFieldValid(board, 1)

--- a/pkg/game/board.go
+++ b/pkg/game/board.go
@@ -252,14 +252,9 @@ func (board *board) fieldCanBePlaced(tile elements.PlacedTile, feat elements.Pla
 
 	// TODO: Field instance cannot expand to another feature on the tile checked by
 	// this method since it's not part of the board. See GH-86.
-	field := field.New(feat, tile.Position)
-	field.Expand(board, board.cityManager)
+	field := field.New(feat, tile)
 
-	// score report function checks the whole field for placed meeples
-	// and reports any meeples that would be returned which we can use here
-	scoreReport := field.GetScoreReport()
-
-	return len(scoreReport.ReturnedMeeples) == 0
+	return field.IsFieldValid(board, 1)
 }
 
 func (board *board) monasteryCanBePlaced(_ elements.PlacedTile, _ elements.PlacedFeature) bool {
@@ -648,7 +643,7 @@ func (board *board) ScoreFinalMeeples() elements.ScoreReport {
 				case feature.Road:
 					miniReport.Join(board.ScoreRoads(pTile, true))
 				case feature.Field:
-					field := field.New(feat, pTile.Position)
+					field := field.New(feat, pTile)
 					field.Expand(board, board.cityManager)
 					miniReport.Join(field.GetScoreReport())
 				case feature.Monastery:

--- a/pkg/game/board_test.go
+++ b/pkg/game/board_test.go
@@ -180,11 +180,9 @@ func TestBoardCanBePlacedReturnsFalseWhenPlacingAtInvalidPosition(t *testing.T) 
 }
 
 func TestBoardFieldCanBePlacedReturnsFalseWhenExpandToFieldWithMeepleHappensOverAnotherField(t *testing.T) {
-	t.Skip("not implemented yet, see GH-86")
-
 	board := NewBoard(tilesets.StandardTileSet()).(*board)
 
-	// prepare board layout (graphical representation can be found in GH-86)
+	// prepare board layout (graphical representation can be found in issue GH-86)
 	tilesToPlace := []elements.PlacedTile{}
 	ptile := elements.ToPlacedTile(tiletemplates.SingleCityEdgeNoRoads().Rotate(2))
 	ptile.Position = position.New(0, 1)

--- a/pkg/game/field/field.go
+++ b/pkg/game/field/field.go
@@ -33,13 +33,21 @@ type Field struct {
 	features           map[fieldKey]struct{}         // this is a set, not a dictionary (the value is always struct{} and only the keys matter)
 	neighbouringCities map[int]struct{}              // this is a set of cities' IDs
 	meeples            []elements.MeepleWithPosition // returned meeples for each player ID (same as in elements.ScoreReport)
+
+	startingTile elements.PlacedTile
 }
 
-func New(feature elements.PlacedFeature, position position.Position) Field {
+func New(feature elements.PlacedFeature, startingTile elements.PlacedTile) Field {
 	features := map[fieldKey]struct{}{
-		{feature: feature, position: position}: {},
+		{feature: feature, position: startingTile.Position}: {},
 	}
-	return Field{features: features, neighbouringCities: map[int]struct{}{}, meeples: []elements.MeepleWithPosition{}}
+	return Field{
+		features:           features,
+		neighbouringCities: map[int]struct{}{},
+		meeples:            []elements.MeepleWithPosition{},
+
+		startingTile: startingTile,
+	}
 }
 
 func (field Field) FeaturesCount() int {
@@ -51,7 +59,13 @@ func (field Field) CitiesCount() int {
 }
 
 // Expands this field to maximum possible size (like flood fill) and finds all neighbouring cities
+// Has to be called on a field which starting tile has already been placed (mostly only at the end of the game)
 func (field *Field) Expand(board elements.Board, cityManager city.Manager) {
+	_, startingTileExists := board.GetTileAt(field.startingTile.Position)
+	if !startingTileExists {
+		panic("the field's starting tile does not exist on the board")
+	}
+
 	newFeatures := map[fieldKey]struct{}{}
 
 	for len(field.features) != 0 {
@@ -61,13 +75,12 @@ func (field *Field) Expand(board elements.Board, cityManager city.Manager) {
 		_, exists := newFeatures[element]
 		if !exists {
 			newFeatures[element] = struct{}{}
-			for _, neighbour := range findNeighbours(element, board) {
+			for _, neighbour := range field.findNeighbours(element, board) {
 				_, exists := newFeatures[neighbour]
 				if !exists {
 					field.features[neighbour] = struct{}{}
 				}
 			}
-
 		}
 
 		// add meeple if it exists
@@ -119,16 +132,51 @@ func (field *Field) Expand(board elements.Board, cityManager city.Manager) {
 	field.features = newFeatures
 }
 
+// Returns true if the field has less than maxMeepleCount meeples, and false otherwise
+func (field Field) IsFieldValid(board elements.Board, maxMeepleCount int) bool {
+	newFeatures := map[fieldKey]struct{}{}
+	meeples := 0
+
+	for len(field.features) != 0 {
+		element, _, _ := utilities.GetAnyElementFromMap(field.features)
+
+		// add neighbouring tiles to the set
+		_, exists := newFeatures[element]
+		if !exists {
+			newFeatures[element] = struct{}{}
+			for _, neighbour := range field.findNeighbours(element, board) {
+				_, exists := newFeatures[neighbour]
+				if !exists {
+					field.features[neighbour] = struct{}{}
+				}
+			}
+		}
+
+		// add meeple if it exists
+		meeple := element.feature.Meeple
+		if meeple.Type != elements.NoneMeeple {
+			meeples++
+			if meeples > maxMeepleCount {
+				return false
+			}
+		}
+
+		// remove the processed field feature from field.features set
+		delete(field.features, element)
+	}
+	return true
+}
+
 // Returns a slice of fieldKey elements containing all features neighbouring a given fieldKey (feature and position)
 // The slice may contain duplicates in some cases (todo?)
-func findNeighbours(field fieldKey, board elements.Board) []fieldKey {
+func (field *Field) findNeighbours(fieldK fieldKey, board elements.Board) []fieldKey {
 	neighbours := []fieldKey{}
 
 	for _, side := range side.EdgeSides {
-		if field.feature.Sides.OverlapsSide(side) {
-			neighbourPosition := field.position.Add(position.FromSide(side))
+		if fieldK.feature.Sides.OverlapsSide(side) {
+			neighbourPosition := fieldK.position.Add(position.FromSide(side))
 
-			tile, tileExists := board.GetTileAt(neighbourPosition)
+			tile, tileExists := field.getTile(neighbourPosition, board)
 			if tileExists {
 				mirroredSide := side.Mirror()
 
@@ -141,6 +189,20 @@ func findNeighbours(field fieldKey, board elements.Board) []fieldKey {
 		}
 	}
 	return neighbours
+}
+
+// If a tile exists in board at the given position, returns board.GetTileAt(position)
+// If the tile doesn't exist in board, but the position is the same as field.startingTile, returns (field.startingTile, true)
+// Otherwise returns board.GetTileAt(position) (that is, empty tile and false)
+func (field Field) getTile(position position.Position, board elements.Board) (elements.PlacedTile, bool) {
+	tile, ok := board.GetTileAt(position)
+	if ok {
+		return tile, ok
+	}
+	if position == field.startingTile.Position {
+		return field.startingTile, true
+	}
+	return tile, ok
 }
 
 // Returns score report for this field. Has to be called after field.Expand() (todo?)

--- a/pkg/game/field/field.go
+++ b/pkg/game/field/field.go
@@ -133,12 +133,22 @@ func (field *Field) Expand(board elements.Board, cityManager city.Manager) {
 }
 
 // Returns true if the field has lesser or equal number of meeples than maxMeepleCount, and false otherwise
+//
+// Useful for testing whether or not a meeple can be placed on a field feature
+// (i.e. are there any other meeples on the expanded field)
+// In such cases, maxMeepleCount should be set to 1 if the tested tile already has a meeple and 0 if it does not.
 func (field Field) IsFieldValid(board elements.Board, maxMeepleCount int) bool {
 	newFeatures := map[fieldKey]struct{}{}
 	meeples := 0
 
-	for len(field.features) != 0 {
-		element, _, _ := utilities.GetAnyElementFromMap(field.features)
+	// copy the original field.features into features, to avoid modifying it
+	features := map[fieldKey]struct{}{}
+	for key := range field.features {
+		features[key] = struct{}{}
+	}
+
+	for len(features) != 0 {
+		element, _, _ := utilities.GetAnyElementFromMap(features)
 
 		// add neighbouring tiles to the set
 		_, exists := newFeatures[element]
@@ -147,7 +157,7 @@ func (field Field) IsFieldValid(board elements.Board, maxMeepleCount int) bool {
 			for _, neighbour := range field.findNeighbours(element, board) {
 				_, exists := newFeatures[neighbour]
 				if !exists {
-					field.features[neighbour] = struct{}{}
+					features[neighbour] = struct{}{}
 				}
 			}
 		}
@@ -161,8 +171,8 @@ func (field Field) IsFieldValid(board elements.Board, maxMeepleCount int) bool {
 			}
 		}
 
-		// remove the processed field feature from field.features set
-		delete(field.features, element)
+		// remove the processed field feature from features set
+		delete(features, element)
 	}
 	return true
 }

--- a/pkg/game/field/field.go
+++ b/pkg/game/field/field.go
@@ -132,7 +132,7 @@ func (field *Field) Expand(board elements.Board, cityManager city.Manager) {
 	field.features = newFeatures
 }
 
-// Returns true if the field has less than maxMeepleCount meeples, and false otherwise
+// Returns true if the field has lesser or equal number of meeples than maxMeepleCount, and false otherwise
 func (field Field) IsFieldValid(board elements.Board, maxMeepleCount int) bool {
 	newFeatures := map[fieldKey]struct{}{}
 	meeples := 0

--- a/pkg/game/field_test.go
+++ b/pkg/game/field_test.go
@@ -70,7 +70,7 @@ func TestScoreFieldOnePlayerGetsPoints(t *testing.T) {
 	}
 
 	// test field.Expand()
-	field := field.New(*tiles[0].GetPlacedFeatureAtSide(side.All, feature.Field), tiles[0].Position)
+	field := field.New(*tiles[0].GetPlacedFeatureAtSide(side.All, feature.Field), tiles[0])
 	field.Expand(board, board.cityManager)
 
 	if field.FeaturesCount() != 12 {
@@ -170,7 +170,7 @@ func TestScoreFieldTwoPlayersGetPoints(t *testing.T) {
 	}
 
 	// test field.Expand()
-	field := field.New(*tiles[0].GetPlacedFeatureAtSide(side.All, feature.Field), tiles[0].Position)
+	field := field.New(*tiles[0].GetPlacedFeatureAtSide(side.All, feature.Field), tiles[0])
 	field.Expand(board, board.cityManager)
 
 	if field.FeaturesCount() != 14 {

